File: sharedModule/src/main/java/ru/nekit/android/qls/setupWizard/BaseSetupWizard.java
package ru.nekit.android.qls.setupWizard;

import android.content.Context;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;

import ru.nekit.android.qls.SettingsStorage;
import ru.nekit.android.qls.pupil.Pupil;
import ru.nekit.android.qls.pupil.PupilManager;
import ru.nekit.android.qls.session.Session;
import ru.nekit.android.qls.session.SessionType;

public abstract class BaseSetupWizard {

    public static final int UNLOCK_SECRET_MIN_SIZE = 4;
    private static final String BIND_CODE_PATTERN = "QLS:%s:%s";
    @NonNull
    protected final Context mContext;
    @NonNull
    protected final PupilManager mPupilManager;
    @NonNull
    protected final SettingsStorage mSettingsStorage;
    @NonNull
    private final List<ISetupWizardStep> mStepStack;
    protected ISetupWizardStep mCurrentStep;

    public BaseSetupWizard(@NonNull Context context) {
        mContext = context;
        mStepStack = new ArrayList<>();
        mSettingsStorage = new SettingsStorage();
        mPupilManager = new PupilManager();
    }

    ISetupWizardStep getNextStep() {
        ISetupWizardStep calculatedStep = calculateNextStep();
        if (needLogin(calculatedStep)) {
            mStepStack.clear();
            mStepStack.add(BaseSetupWizardStep.UNLOCK_SECRET);
        }
        if (calculatedStep.needInternetConnection()) {
            if (!internetIsConnected()) {
                mStepStack.add(BaseSetupWizardStep.SETUP_INTERNET_CONNECTION);
            }
        }
        int size = mStepStack.size();
        if (size == 0 || (size > 0 && mStepStack.get(size - 1) != calculatedStep)) {
            mStepStack.add(calculatedStep);
        }
        return mStepStack.remove(0);
    }

    private boolean internetIsConnected() {
        return true;
    }

    @NonNull
    protected abstract ISetupWizardStep calculateNextStep();

    public void setIntroductionIsPresented(boolean value) {
        mSettingsStorage.setIntroductionIsPresented(value);
    }

    protected boolean setupIsCompleted() {
        return mSettingsStorage.setupWizardIsCompleted(getName());
    }

    public boolean setupIsStarted() {
        return mSettingsStorage.setupWizardIsStarted(getName());
    }

    protected abstract String getName();

    public int getQuestSeriesLength() {
        return mSettingsStorage.getQuestSeriesLength();
    }

    public void setQuestSeriesLength(int value) {
        mSettingsStorage.setQuestSeriesLength(value);
    }

    protected void startSetupWizard() {
        mSettingsStorage.startSetupWizard(getName());
    }

    protected void startSetupWizard(boolean value) {
        mSettingsStorage.startSetupWizard(getName(), value);
    }

    protected void completeSetupWizard() {
        mSettingsStorage.completeSetupWizard(getName());
    }

    protected void completeSetupWizard(boolean value) {
        mSettingsStorage.completeSetupWizard(getName(), value);
    }

    @NonNull
    public String createBindCode() {
        Pupil pupil = mPupilManager.getCurrentPupil();
        assert pupil != null;
        return String.format(BIND_CODE_PATTERN, pupil.getUuid(), pupil.name);
    }

    @NonNull
    protected Pupil createPupilFromBindCode(@NonNull String bindCode) {
        String[] values = TextUtils.split(bindCode, ":");
        Pupil pupil = new Pupil(values[1]);
        pupil.name = values[2];
        return pupil;
    }

    boolean needLogin(@NonNull ISetupWizardStep step) {
        return step.needLogin() && !Session.isValid(SessionType.SETUP_WIZARD);
    }

    ISetupWizardStep getCurrentStep() {
        return mCurrentStep;
    }
}

File: sharedModule/src/main/java/ru/nekit/android/qls/setupWizard/BaseSetupWizard.kt
package ru.nekit.android.qls.setupWizard

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.support.v4.content.ContextCompat
import android.text.TextUtils
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import ru.nekit.android.qls.shared.model.Pupil

abstract class BaseSetupWizard(protected val context: Context) {

    private val stepStack: MutableList<ISetupWizardStep>

    init {
        stepStack = ArrayList()
    }

    var currentStep: ISetupWizardStep? = null

    internal val nextStep: Single<ISetupWizardStep>
        get() = calculateNextStep().flatMap { step ->
            Single.zip(needLogin(step), Single.just(step.needInternetConnection() && !internetIsConnected()),
                    BiFunction { needLogin: Boolean, needInternetConnection: Boolean ->
                        if (needLogin) {
                            stepStack.clear()
                            stepStack.add(BaseSetupWizardStep.UNLOCK_SECRET)
                        }
                        if (needInternetConnection) {
                            stepStack.add(BaseSetupWizardStep.SETUP_INTERNET_CONNECTION)
                        }
                        step
                    }).map {
                val size = stepStack.size
                if (size == 0 || size > 0 && stepStack[size - 1] !== it) {
                    stepStack.add(it)
                }
                stepStack.removeAt(0)
            }
        }

    private fun internetIsConnected(): Boolean {
        return true
    }

    protected abstract fun calculateNextStep(): Single<ISetupWizardStep>

    abstract fun setupIsStart(body: (Boolean) -> Unit)

    abstract fun setupIsComplete(body: (Boolean) -> Unit)

    abstract fun startSetupWizard()

    abstract fun completeSetupWizard()

    protected fun createPupilFromBindCode(bindCode: String): Pupil {
        val values = TextUtils.split(bindCode, ":")
        return Pupil(values[1], values[2])
    }

    abstract fun needLogin(step: ISetupWizardStep): Single<Boolean>

    private fun permissionIsGranted(permission: String): Boolean = Build.VERSION.SDK_INT < Build.VERSION_CODES.M ||
            ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED

    fun permissionIsGranted(permissions: Array<String>): Boolean =
            permissions.map { permissionIsGranted(it) }.all { it }

    fun callPhonePermissionIsGranted(): Boolean =
            permissionIsGranted(Manifest.permission.CALL_PHONE)

    fun readContactsPermissionIsGranted(): Boolean =
            permissionIsGranted(Manifest.permission.READ_CONTACTS)

    companion object {

        const val UNLOCK_SECRET_MIN_SIZE = 4
        const val BIND_CODE_PATTERN = "QLS:%s:%s"
    }

}